README.md
# Flutter Termii

<p align="left">
    <img title="Termii" src="https://termii.com/assets/images/logo.png"/>
</p>

[![pub package](https://img.shields.io/pub/v/flutter_termii.svg)](https://pub.dev/packages/flutter_termii)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://opensource.org/licenses/MIT)
[![style: very good analysis](https://img.shields.io/badge/style-very_good_analysis-B22C89.svg)](https://pub.dev/packages/flutter_termii)

---

A Flutter plugin that helps developers use messaging channels to verify and authenticate customer transactions

|             | Android | iOS  | Linux | macOS  | Web | Windows     |
|-------------|---------|------|-------|--------|-----|-------------|
| **Support** |   Any   | Any  |  Any  |  Any   | Any |   Any       |

## Features

Use this plugin in your Flutter app to:

- Send message
- Send Bulk message
- Send Token
- Send Voice Token
- Voice Call
- Generate In-App Token
- Verify Token
- View Balance

## Getting started

In order to use Termii's APIs, you need to first create an account for free at [www.termii.com](https://termii.com/).

### BASE URL

Your Termii account has its own Base URL, which you should use in all API requests.
The base URL shown below is a sample base URL. Your base URL can be found on your [dashboard](https://accounts.termii.com/#/).

### Retrieving your API Key

Your API key can be obtained from your [dashboard settings](https://accounts.termii.com/#/account/api).

## Usage

To use this plugin, add `flutter_termii` as a [dependency in your pubspec.yaml file](https://flutter.dev/platform-plugins/)

```yaml
dependencies:
  flutter:
    sdk: flutter
  flutter_termii:
```

```dart
import 'package:flutter/material.dart';
import 'package:flutter_termii/flutter_termii.dart';

final termii = Termii(
  url: 'https://api.ng.termii.com',
  apiKey: 'YOUR API KEY',
  senderId: 'CompanyName',
);

final responseData = await termii.sendSms(
  destination: "2349012672787",
  message: "This is a test message",
);

print(responseData);
```

## Credits

Developed with 💙 by [Abdulazeez Oshinowo](https://github.com/Oshinowo/)

## License

The MIT License (MIT). Please see [License File](https://github.com/Oshinowo/flutter_termii/blob/main/LICENSE) for more information.

## Additional information

Termii helps deliver great customer messaging experience. You can dive deep into the [full API Reference Documentation](https://developers.termii.com/) to seamlessly integrate it's messaging channels and verification functionalities into your product.


CHANGELOG.md
## 0.2.1

* Update License.

## 0.2.0

* update sdk.

## 0.1.9

* fixes.

## 0.1.7

* platforms update.

## 0.1.6

* fixes.

## 0.1.5

* Udate documentation.

## 0.1.4

* Update License.

## 0.1.2

* Update ReadMe.

## 0.1.1

* Add documentation.

## 0.0.1

* Initial Open Source release.


lib/flutter_termii.dart
library flutter_termii;

import 'src/network_helper.dart';
import 'dart:convert';

/// An API to send and verify text/voice messages.
class Termii {
  /// The base URL is used to route your request to the appropriate "regulatory region" and to optimize traffic between data centers with the region.
  final String url;

  /// API calls are authenticated by including your API key in the body of the request you make.
  final String apiKey;

  /// A Sender ID is the name or number that identifies the sender of an SMS message.
  final String senderId;

  Termii({
    required this.url,
    required this.apiKey,
    required this.senderId,
  });

  /// Send text messages across different messaging channels.
  Future<dynamic> sendSms({
    /// Represents the destination phone number. Phone number must be in the international format (Example: 23490126727).
    required String destination,

    /// Text of a message that would be sent to the destination phone number.
    required String message,

    /// This is the route through which the message is sent. It is either dnd, whatsapp, or generic.
    String channel = 'generic',
  }) async {
    final response = await NetworkHelper.postRequest(
      url: "$url/api/sms/send",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'to': destination,
        'from': senderId,
        'sms': message,
        'type': 'plain',
        'channel': channel,
        'api_key': apiKey,
      }),
    );
    return response.body;
  }

  /// Send media messages across different messaging channels.
  Future<dynamic> sendMediaSms({
    /// Represents the destination phone number. Phone number must be in the international format (Example: 23490126727).
    required String destination,

    /// The url to the file resource.
    required String mediaUrl,

    /// This is the route through which the message is sent. It is either dnd, whatsapp, or generic.
    String channel = 'generic',

    /// The caption that should be added to the image,
    String? mediaCaption,
  }) async {
    final response = await NetworkHelper.postRequest(
      url: "$url/api/sms/send",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'to': destination,
        'from': senderId,
        'type': 'plain',
        'channel': channel,
        'api_key': apiKey,
        'media': {
          'url': mediaUrl,
          'caption': mediaCaption,
        }
      }),
    );
    return response.body;
  }

  /// Send bulk text messages across different messaging channels to multiple phone numbers.
  Future<dynamic> sendBulkSms({
    /// Represents the array of phone numbers you are sending to (Example: ["23490555546", "23423490126999","23490555546"]).
    required List<String> multipleDestinations,

    /// Text of a message that would be sent to the destination phone number.
    required String message,
    String channel = 'generic',
  }) async {
    final response = await NetworkHelper.postRequest(
      url: "$url/api/sms/send/bulk",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'to': multipleDestinations,
        'from': senderId,
        'sms': message,
        'type': 'plain',
        'channel': channel,
        'api_key': apiKey,
      }),
    );
    return response.body;
  }

  /// Send messages with auto-generated messaging numbers that adapt to customers location.
  Future<dynamic> sendSmsWithAutoGeneratedNumber({
    required String destination,
    required String message,
  }) async {
    final response = await NetworkHelper.postRequest(
      url: "$url/api/sms/number/send",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'to': destination,
        'sms': message,
        'api_key': apiKey,
      }),
    );
    return response.body;
  }

  /// The send token triggers one-time-passwords (OTP) across any available messaging channel.
  Future<dynamic> sendToken({
    /// The destination phone number. Phone number must be in the international format (Example: 23490126727).
    required String destination,

    /// Type of message that will be generated and sent as part of the OTP message. You can set message type to numeric or alphanumeric.
    required MessageType messageType,
    String channel = 'generic',
    required int pinAttempts,

    /// Represents how long the PIN is valid before expiration. The time is in minutes. The minimum time value is 0 and the maximum time value is 60.
    required int pinExpiryTime,

    /// The length of the PIN code.It has a minimum of 4 and maximum of 8.
    required int pinLength,

    /// Example: "< 1234 >".
    ///
    /// PIN placeholder. Right before sending the message, PIN code placeholder will be replaced with generate PIN code.
    required String pinPlaceholder,

    /// Text of a message that would be sent to the destination phone number.
    required String messageText,

    /// Type of PIN code that will be generated and sent as part of the OTP message. It has a minimum of one attempt.
    required MessageType pinType,
  }) async {
    final response = await NetworkHelper.postRequest(
      url: "$url/api/sms/otp/send",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'api_key': apiKey,
        'message_type':
            messageType == MessageType.numeric ? 'NUMERIC' : 'ALPHANUMERIC',
        'to': destination,
        'from': senderId,
        'channel': channel,
        'pin_attempts': pinAttempts,
        'pin_time_to_live': pinExpiryTime,
        'pin_length': pinLength,
        'pin_placeholder': pinPlaceholder,
        'message_text': messageText,
        'pin_type': pinType == MessageType.numeric ? 'NUMERIC' : 'ALPHANUMERIC',
      }),
    );
    return response.body;
  }

  /// Trigger one-time passwords (OTP) through the voice channel to a phone number.
  Future<dynamic> sendVoiceToken({
    required String destination,
    required int pinAttempts,
    required int pinExpiryTime,
    required int pinLength,
  }) async {
    final response = await NetworkHelper.postRequest(
      url: "$url/api/sms/otp/send/voice",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'api_key': apiKey,
        'phone_number': destination,
        'pin_attempts': pinAttempts,
        'pin_time_to_live': pinExpiryTime,
        'pin_length': pinLength,
      }),
    );
    return response.body;
  }

  /// Send messages from your application through our voice channel to a phone number.
  Future<dynamic> sendVoiceCall({
    required String destination,
    required int code,
  }) async {
    final response = await NetworkHelper.postRequest(
      url: "$url/api/sms/otp/call",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'api_key': apiKey,
        'phone_number': destination,
        'code': code,
      }),
    );
    return response.body;
  }

  /// This API returns OTP codes in JSON format which can be used within any web or mobile app.
  Future<dynamic> generateInAppToken({
    required String destination,
    required int pinAttempts,
    required int pinExpiryTime,
    required int pinLength,
    required MessageType pinType,
  }) async {
    final response = await NetworkHelper.postRequest(
      url: "$url/api/sms/otp/generate",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'api_key': apiKey,
        'phone_number': destination,
        'pin_attempts': pinAttempts,
        'pin_time_to_live': pinExpiryTime,
        'pin_length': pinLength,
        'pin_type': pinType == MessageType.numeric ? 'NUMERIC' : 'ALPHANUMERIC',
      }),
    );
    return response.body;
  }

  /// Verify token API, checks tokens sent to customers and returns a response confirming the status of the token.
  Future<dynamic> verifyToken({
    required String pinId,
    required String pin,
  }) async {
    final response = await NetworkHelper.postRequest(
      url: "$url/api/sms/otp/verify",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, dynamic>{
        'api_key': apiKey,
        'pin_id': pinId,
        'pin': pin,
      }),
    );
    return response.body;
  }

  /// The Balance API returns your total balance and balance information from your wallet, such as currency.
  Future<dynamic> veiwBalance({
    required String pinId,
    required String pin,
  }) async {
    final response = await NetworkHelper.getRequest(
      url: "$url/api/get-balance?api_key=$apiKey",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
    );
    return response;
  }

  /// The search API allows businesses verify phone numbers and automatically detect their status as well as current network.
  Future<dynamic> searchPhoneNumber({
    required String phoneNumber,
  }) async {
    final response = await NetworkHelper.getRequest(
      url: "$url/api/check/dnd?api_key=$apiKey&phone_number=$phoneNumber",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
    );
    return response;
  }

  /// The status API allows businesses to detect if a number is fake or has ported to a new network.
  Future<dynamic> detectPhoneNumberStatus({
    required String phoneNumber,
    required String countryCode,
  }) async {
    final response = await NetworkHelper.getRequest(
      url:
          "$url/api/insight/number/query?phone_number=$phoneNumber&api_key=$apiKey&country_code=$countryCode",
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
    );
    return response;
  }
}

enum MessageType { numeric, alphanumeric }


example/example.dart
import 'package:flutter/material.dart';
import 'package:flutter_termii/flutter_termii.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Flutter Termii',
      debugShowCheckedModeBanner: false,
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final _formKey = GlobalKey<FormState>();
  final _phonenumberController = TextEditingController();

  // Authenticate requests with valid Base URL and API Key
  final termii = Termii(
    url: 'https://api.ng.termii.com',
    apiKey: 'YOUR API KEY',
    senderId: 'CompanyName',
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Form(
          key: _formKey,
          child: Padding(
            padding: const EdgeInsets.all(10.0),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                TextFormField(
                  controller: _phonenumberController,
                  decoration: const InputDecoration(
                    contentPadding: EdgeInsets.symmetric(
                      vertical: 8.0,
                      horizontal: 12.0,
                    ),
                    border: OutlineInputBorder(),
                  ),
                ),
                const SizedBox(
                  height: 20,
                ),
                SizedBox(
                  width: double.infinity,
                  height: 60,
                  child: TextButton(
                    onPressed: _submit,
                    style: TextButton.styleFrom(
                      backgroundColor: Colors.blue,
                      primary: Colors.white,
                    ),
                    child: const Text('Submit'),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  _submit() async {
    // destination/PhoneNumber must be in the international format (Example: 23490126727).
    final responseData = await termii.sendSms(
      destination: _phonenumberController.text,
      message: "This is working",
    );

    final bulkResponseData = await termii.sendBulkSms(
      multipleDestinations: ["23490555546", "23423490126999", "23490555546"],
      message: "This is working for multiple numbers",
      // channel is the route through which the message is sent. It is either dnd, whatsapp, or generic (default)
      channel: "whatsapp",
    );

    final tokenResponseData = await termii.sendToken(
      destination: _phonenumberController.text,
      pinAttempts: 2, // 2 times
      // pinExpiryTime represents how long the PIN is valid before expiration. The time is in minutes. The minimum time value is 0 and the maximum time value is 60
      pinExpiryTime: 2, // 2 minutes
      pinLength: 6,
      // Right before sending the message, PIN code placeholder will be replaced with generated PIN code.
      pinPlaceholder: "< 1234 >",
      messageText: "Your pin is < 1234 >",
      messageType: MessageType.numeric,
      pinType: MessageType.numeric,
    );

    final inAppTokenResponseData = await termii.generateInAppToken(
      destination: _phonenumberController.text, // Example: 2349012672789
      pinAttempts: 6, // 6 times
      pinExpiryTime: 5, // 5 minutes
      pinLength: 3,
      pinType: MessageType.numeric,
    );

    final verifyTokenResponseData = await termii.verifyToken(
      pinId: "c8dcd048-5e7f-4347-8c89-4470c3af0b",
      pin: "195558",
    );

    final searchResponseData = await termii.searchPhoneNumber(
      phoneNumber: _phonenumberController.text,
    );

    print(responseData);
  }
}
